sdk/core/core/src/az_span.c

```diff
@@ -279,7 +279,7 @@ void az_span_to_str(char* destination, int32_t destination_max_size, az_span sou
   // Implementations of memmove generally do the right thing when number of bytes to move is 0, even
   // if the ptr is null, but given the behavior is documented to be undefined, we disallow it as a
   // precondition.
-  AZ_PRECONDITION_VALID_SPAN(source, 0, false);
+  AZ_PRECONDITION_VALID_SPAN(source, 0);
```

Comment:
Most APIs like these that rejected a "null span" (but allowed an empty span), need to retain this behavior. We can't let users pass in `AZ_SPAN_NULL` here.

This is especially true for methods that call underlying STL APIs like memmove which are documented to have undefined behavior for null ptr. With the precondition, the caller knows at dev time not to pass in a "null span".

The problem with this change is that it forces us to turn a precondition check to a "real" check to guard against null ptr.

If you want to separate the "null_is_valid" check into a separate precondition so that all callers don't have to pass in true/false, that is fine, but I don't think we should be changing the behavior of APIs.

That said, since this is preconditions, saving a bool is not worth it, since that cost shouldn't affect the user in production anyway.

Alternatively, we should change the precondition to always reject a span with `null` ptr.

Reply:
Do you want it to look like this? - https://github.com/Azure/azure-sdk-for-c/pull/644/files

Reply:
I don't follow why to exclude spans with { .ptr = NULL, .size = 0 } in all the places across our SDK (with single exception for az_span_slice), just because memmove() from standard library is not formally documented for the case when size=0 ptr=NULL, and that case is enclosed in az_span_to_str() function, while the rest of our SDK would work correctly handling spans of size 0 reagrdless of the pointer. Even if you get a ptr, and pass it to a function, and it imposes the limitation that ptr can't be NULL even when size is 0, then we should handle that in the code that invokes that function, and I am not even sure it should be a precondition then, and not a simple if(). I don't think we should impose the limitations on the all the spans across all our code, just because of one external function.
We only use memmove in az_span_to_str() and in az_span_copy(). The rest of the code is using az_span_copy() when it needs to memmove. And BTW, az_span_copy hadles it with `if(size == 0) return destination`. I think az_span_to_str() should have the same sort of candling for size == 0 case.